### PR TITLE
Uniswap v2 sqlite update

### PR DESCRIPTION
**Pull request by jpsains**
* Insert to db statement updated to SQL transaction
* Uniswap V2 adapter now uses factory and graphQl
* Ive dropped a couple of onchain requests from factory implementation e.g. getSupply and getProtocolTokenMetadata for the protocol token which significantly improves performance 
* Implemented p-queue 
* Created a script npm run check-metadata-type
* Created a script to check total pools per adapter in the db

Performance:
* 10k tokens takes 52 seconds to load
* Approximately taking 30mins to load 300k tokens

Note: I got this warning when pushing the ethereum db 
```
remote: warning: File ethereum.db is 75.95 MB; this is larger than GitHub's recommended maximum file size of 50.00 MB
remote: warning: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
```

### CODE REVIEW

File: packages/adapters-library/src/config.ts [hunk 1]
```diff
@@ -122,7 +122,7 @@ const ConfigSchema = z
           .default(
             parseNumberEnv(
               process.env.DEFI_ADAPTERS_MAX_BATCH_SIZE_ETHEREUM,
-              100,
+              1000,
```
Comment (jpsains) [inline packages/adapters-library/src/config.ts:125]: Happy to change this but 1000 works pretty well imo

File: packages/adapters-library/src/scripts/index.ts [hunk 2]
```diff
@@ -22,6 +23,8 @@ const adaptersController = defiProvider.adaptersController
 
 featureCommands(program, defiProvider)
 
+checkMetadataType(program, chainProviders, adaptersController)
```
Comment (jpsains) [inline packages/adapters-library/src/scripts/index.ts:26]: New script 

```
Database Cached Adapters:
  Protocol ID: aave-v2 - Product ID: a-token
  Protocol ID: aave-v2 - Product ID: stable-debt-token
  Protocol ID: aave-v2 - Product ID: variable-debt-token
  Protocol ID: aave-v3 - Product ID: a-token
  Protocol ID: aave-v3 - Product ID: stable-debt-token
  Protocol ID: aave-v3 - Product ID: variable-debt-token
  Protocol ID: chimp-exchange - Product ID: pool
  Protocol ID: convex - Product ID: pool
  Protocol ID: convex - Product ID: staking
  Protocol ID: convex - Product ID: cvxcrv-wrapper
  Protocol ID: convex - Product ID: sidechain-staking
  Protocol ID: etherfi - Product ID: e-eth
  Protocol ID: etherfi - Product ID: we-eth
  Protocol ID: etherfi - Product ID: liquid
  Protocol ID: etherfi - Product ID: l2
  Protocol ID: gmx - Product ID: glp
  Protocol ID: lido - Product ID: st-eth
  Protocol ID: lido - Product ID: wst-eth
  Protocol ID: lynex - Product ID: classic
  Protocol ID: maker - Product ID: s-dai
  Protocol ID: mendi-finance - Product ID: supply-market
  Protocol ID: mendi-finance - Product ID: borrow-market
  Protocol ID: morpho-blue - Product ID: vault
  Protocol ID: pancakeswap-v2 - Product ID: pool
  Protocol ID: rocket-pool - Product ID: reth
  Protocol ID: spark-v1 - Product ID: sp-token
  Protocol ID: spark-v1 - Product ID: variable-debt-token
  Protocol ID: stakewise - Product ID: os-eth
  Protocol ID: stargate - Product ID: pool
  Protocol ID: stargate - Product ID: farm
  Protocol ID: stargate - Product ID: pool-v2
  Protocol ID: stargate - Product ID: farm-v2
  Protocol ID: sushiswap-v2 - Product ID: pool
  Protocol ID: swell - Product ID: sw-eth
  Protocol ID: uniswap-v2 - Product ID: pool
  Protocol ID: xfai - Product ID: dex
  Total DB Adapters: 36

File Cached Adapters:
  Protocol ID: angle-protocol - Product ID: savings
  Protocol ID: beefy - Product ID: cow-token
  Protocol ID: beefy - Product ID: moo-token
  Protocol ID: beefy - Product ID: rcow-token
  Protocol ID: beefy - Product ID: rmoo-token
  Protocol ID: compound-v2 - Product ID: supply-market
  Protocol ID: compound-v2 - Product ID: borrow-market
  Protocol ID: curve - Product ID: pool
  Protocol ID: curve - Product ID: staking
  Protocol ID: curve - Product ID: voting-escrow
  Protocol ID: deri - Product ID: pool
  Protocol ID: ethena - Product ID: ethena
  Protocol ID: flux - Product ID: supply-market
  Protocol ID: flux - Product ID: borrow-market
  Protocol ID: morpho-aave-v2 - Product ID: optimizer-borrow
  Protocol ID: morpho-aave-v2 - Product ID: optimizer-supply
  Protocol ID: morpho-aave-v3 - Product ID: optimizer-supply
  Protocol ID: morpho-aave-v3 - Product ID: optimizer-borrow
  Protocol ID: morpho-blue - Product ID: market-supply
  Protocol ID: morpho-blue - Product ID: market-borrow
  Protocol ID: morpho-compound-v2 - Product ID: optimizer-supply
  Protocol ID: morpho-compound-v2 - Product ID: optimizer-borrow
  Protocol ID: pendle - Product ID: yield-token
  Protocol ID: pendle - Product ID: principle-token
  Protocol ID: pendle - Product ID: standardised-yield-token
  Protocol ID: pendle - Product ID: lp-token
  Protocol ID: quickswap-v2 - Product ID: pool
  Protocol ID: renzo - Product ID: ez-eth
  Protocol ID: solv - Product ID: solv-btc
  Protocol ID: sonne - Product ID: supply-market
  Protocol ID: sonne - Product ID: borrow-market
  Protocol ID: stargate - Product ID: voting-escrow
  Total File Adapters: 32

No pool cache adapters:
  Protocol ID: carbon-defi - Product ID: strategies
  Protocol ID: iziswap - Product ID: pool
  Protocol ID: lynex - Product ID: algebra
  Protocol ID: prices-v2 - Product ID: usd
  Protocol ID: solv - Product ID: yield-market
  Protocol ID: syncswap - Product ID: pool
  Protocol ID: uniswap-v3 - Product ID: pool
  Total Other Adapters: 7
```

File: packages/adapters-library/src/scripts/helpers.ts [hunk 3]
```diff
@@ -458,8 +458,8 @@ export class Helpers {
       return onChainTokenMetadata
     }
 
-    const errorMessage = 'Cannot find token metadata for token'
-    logger.error({ tokenAddress, chainId: this.chainId }, errorMessage)
```
Comment (jpsains) [inline packages/adapters-library/src/scripts/helpers.ts:462]: including the error object was flooding terminal

File: packages/adapters-library/src/scripts/buildMetadata.ts [hunk 4]
```diff
@@ -81,28 +81,6 @@ export function buildMetadata(
...
-                    fileDetails: {
-                      protocolId: Protocol
-                      productId: string
-                      chainId: Chain
-                      fileKey: string
-                    }
-                  }
-                | undefined
-            }
-
```
Comment (jpsains) [inline packages/adapters-library/src/scripts/buildMetadata.ts:105]: Removing this temporary solution that cached the new metadata shape into a file

File: packages/adapters-library/src/core/decorators/cacheToFile.ts [hunk 5]
```diff
@@ -66,6 +66,7 @@ export function CacheToFile({ fileKey }: { fileKey: string }) {
 
       return metadata
     }
+    replacementMethod.isCacheToFileDecorated = true
```
Comment (jpsains) [inline packages/adapters-library/src/core/decorators/cacheToFile.ts:69]: So I can report the metadata type

File: packages/adapters-library/src/core/adaptersController.ts [hunk 6]
```diff
@@ -156,12 +157,25 @@ export class AdaptersController {
...
+          existingAdapter.getProtocolDetails().protocolId
+        } ${existingAdapter.getProtocolDetails().productId} ${
+          existingAdapter.getProtocolDetails().chainId
+        } and is duplicated by ${adapter.getProtocolDetails().protocolId} ${
+          adapter.getProtocolDetails().productId
+        } ${adapter.getProtocolDetails().chainId}`
+
+        logger.error(errorMessage)
+
+        throw new Error(errorMessage)
```
Comment (jpsains) [inline packages/adapters-library/src/core/adaptersController.ts:178]: Looks messy but its a life saver....I had some duplicate pools in the db

File: packages/adapters-library/src/SQLiteMetadataProvider.ts [hunk 7]
```diff
@@ -12,6 +12,8 @@ export type IMetadataProvider = {
     protocolId: Protocol
     productId: string
   }) => Promise<ProtocolToken[]>
+
+  getPoolCount: (protocolId: string, productId: string) => Promise<number>
```
Comment (jpsains) [inline packages/adapters-library/src/SQLiteMetadataProvider.ts:16]: New method to get pool count from db per adapter

File: packages/adapters-library/src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts [hunk 8]
```diff
@@ -22,7 +23,6 @@ export class AaveV3ATokenPoolAdapter extends AaveBasePoolAdapter {
   adapterSettings = {
     enablePositionDetectionByProtocolTokenTransfer: true,
     includeInUnwrap: true,
-    version: 2,
```
Comment (jpsains) [inline packages/adapters-library/src/adapters/aave-v3/products/a-token/aaveV3ATokenAdapter.ts:25]: Dont need this differentiation as the declarators now add new value to the getProtocolTokens and buildMetadata methods

File: packages/adapters-library/src/core/adapters/UniswapV2PoolForkAdapter.ts [hunk 9]
```diff
@@ -435,143 +471,267 @@ export abstract class UniswapV2PoolForkAdapter
...
-              id
-              token0 {
-                id
-              }
-              token1 {
+              pairs(
+                first: ${BATCH_SIZE}
+                skip: ${skip}
+                where: {
+                  reserveUSD_gt: 50000
```
Comment (jpsains) [inline packages/adapters-library/src/core/adapters/UniswapV2PoolForkAdapter.ts:547]: Get pools with more than $50k if graphQL path selected

File: packages/adapters-library/src/scripts/helpers.ts [hunk 10]
```diff
@@ -60,6 +60,13 @@ export class Helpers {
   }: GetPositionsInput & {
     protocolTokens: Erc20Metadata[]
   }): Promise<ProtocolPosition[]> {
+    // Otherwise we might overload the node
+    if (!protocolTokenAddresses && protocolTokens.length > 1000) {
+      throw new Error(
+        'Too many tokens to fetch, protocolTokenAddresses must be provided',
+      )
+    }
+
```
Comment (jpsains) [inline packages/adapters-library/src/scripts/helpers.ts:69]: Bit of protection here....now that we can support more pools bit worries we might try and get thousands of pools

File: packages/adapters-library/src/core/utils/getTokenMetadata.ts [hunk 11]
```diff
@@ -52,8 +52,8 @@ export async function getTokenMetadata(
     return onChainTokenMetadata
   }
 
-  const errorMessage = 'Cannot find token metadata for token'
-  logger.error({ tokenAddress, chainId }, errorMessage)
+  const errorMessage = 'Failed getting token metadata'
+  logger.error({ tokenAddress, chainId })
```
Comment (bergarces) [inline packages/adapters-library/src/core/utils/getTokenMetadata.ts:56]: It'll be difficult to know what error it is in the logs with just the object, it should also include the message.

File: packages/adapters-library/src/defiProvider.ts [hunk 12]
```diff
@@ -61,14 +61,23 @@ function buildMetadataProviders(): Record<Chain, IMetadataProvider> {
 const dbParams = (chainId: Chain): [string, Database.Options] => {
   const dbPath = path.join(__dirname, '../../..', `${ChainName[chainId]}.db`)
 
-  if (!existsSync(dbPath)) {
+  if (
+    !(process.env.DEFI_ALLOW_DB_CREATION !== 'false') &&
+    !existsSync(dbPath)
+  ) {
```
Comment (bergarces) [inline packages/adapters-library/src/defiProvider.ts:67]: In what event returning a `dbPath` that doesn't exist will work?

Comment (jpsains) [reply packages/adapters-library/src/defiProvider.ts:67]: If database creation is not allowed and the database file doesn't exist, it'll throw

**Pull request merged**